pkg/apis/eventing/v1alpha1/provisioner_types.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"encoding/json"
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Provisioner provisions.
+type Provisioner struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ProvisionerSpec   `json:"spec"`
+	Status ProvisionerStatus `json:"status"`
+}
+
+// Check that Provisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*Provisioner)(nil)
+var _ apis.Defaultable = (*Provisioner)(nil)
+var _ runtime.Object = (*Provisioner)(nil)
+var _ webhook.GenericCRD = (*Provisioner)(nil)
+
+// ProvisionerSpec is the spec for a Provisioner resource.
+type ProvisionerSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Type is the type of the resource to be provisioned.
+	// +required
+	Type runtime.TypeMeta `json:"type"`
```

Comment:
Side comment, non-blocking:

`TypeMeta` typically contains a specific API version (e.g. `eventing.knative.dev/v1alpha1`), but seems like it should be a non-versioned "api group" plus a kind. `TypeMeta` seems close enough, though.

Reply:
Once CRD conversion is supported, we might not _need_ version, but until then provisioners probably need to specific which version they support.

Reply:
IIUC, the intent is that the different API versions (e.g. v1alpha1, v1beta1, etc) all represent the same underlying object.